Card/src/main/java/com/paypal/android/card/CardResultCallback.java

```diff
@@ -0,0 +1,7 @@
+package com.paypal.android.card;
+
+import androidx.annotation.NonNull;
+
+public interface CardResultCallback {
```

Comment:
Can we convert this interface to Kotlin? I believe the Kotlin equivalent will still work for Java integrations.

```kotlin
interface CardResultCallback {
    fun onCardResult(result: CardResult)
}
```

Reply:
@tdchow @andrespelaezp I agree that a pure Kotlin codebase is appealing, but if we do that, we lose the function-like interface in Kotlin:

```
cardClient.approveOrder(orderId, card, object : CardResultCallback {
  override fun onCardResult(result: CardResult) {
    when (result) {
      is CardResult.Success -> {
        // handle success
      }
      is CardResult.Error -> {
        // handle error
      }
    }
  }
})
```

Reply:
Java looks great with a Kotlin interface though, but Kotlin would take a hit if we use non-Java callbacks. I thinks that's the reverse of what we want. Java-style callbacks give us the best of both worlds.

Reply:
Maybe we can put together what all 4 of the code samples would look like. So 2 snippets (Java & Kotlin) for how this looks for a merchant if the source code is written in Java, and 2 snippets (Java & Kotlin) for how this looks for a merchant if the source code is written in Kotlin.

This would help, me at least, visualize our options.

Reply:
Correct me if I'm wrong, but it sounds like what Steven is advocated for is to not require our Java merchants to do a cast like `if (result instanceof CardResult.Success)`. So on one hand our Kotlin integration looks super slick - but our Java integration is clunky. This proposes a middle ground? Again - I think the explicit snippets next to each other would help visualize.

Reply:
> Java looks great with a Kotlin interface though, but Kotlin would take a hit if we use non-Java callbacks. I thinks that's the reverse of what we want. Java-style callbacks give us the best of both worlds.

Ah right, I remember running into this while we were working on the 3P NXO SDK. I think the only caveat with going with a Java defined interface is that the Kotlin call site requires the interface type to be defined.

```kotlin
CardClient().approveOrder { data ->

}

// vs

CardClient().approveOrder(JavaInterface { data ->

})
```

Here's what we did in the NXO SDK to get around the issue:

```kotlin
class CardClient {

    fun approveOrder(testInterface: TestInterface) {
        testInterface.testFunction("")
    }

    fun approveOrder(callback: (String) -> Unit) {
        approveOrder(object : TestInterface {
            override fun testFunction(field: String) {
                callback.invoke(field)
            }
        })
    }
}
```
We have multiple functions on the object to satisfy both languages. One of the issues with this pattern is that a merchant could call the wrong function for their specific language. 

I came across this section in the Kotlin-Java interop guide in the developer docs that calls out this exact case: https://developer.android.com/kotlin/interop#lambda_arguments To Sammy's point, I think it would help to have integration examples for each of the solutions.

Reply:
It actually compiles as-is right now. The `CardViewModel` code is unchanged [here](https://github.com/paypal/Android-SDK/blob/3d73756b6000ee1298f4929e39c22be197e30d67/Demo/src/main/java/com/paypal/android/ui/card/CardViewModel.kt#L81). I have seen that style before, but in this case, it looks like Kotlin perceives a Function style callback.

Reply:
Ah, you're right! Looks like this was added in Kotlin 1.4: https://kotlinlang.org/docs/whatsnew14.html#sam-conversions-for-kotlin-interfaces 

I was originally testing with an older version of Kotlin.